### ISSUE TITLE
Automatically stop scanning if in area where reports have been created

### ISSUE THREAD

**Issue by mjaakko** [mjaakko/NeoStumbler#11]
Currently scanning can be started automatically when the device is moving in an area where no reports have been created recently. Scanning will continue until the device stops moving. The user should be able to toggle whether the scanning should stop automatically if the device moves to an area where reports have been created recently and the scanning was started automatically

**Comment by lucasmz-dev** [comment]
This would be definitely quite nice. I mostly stay at fixed places, and NeoStumbler stays on for areas where it has more than enough data, and actually, I noticed that because of my device's GPS accuracy I suppose, it starts to mark more areas as scanned where I'm not near them.

I mean, I *could* toggle it manually when I'm out and about, but I'll most likely forget it about it.

**Comment by Mikaela** [comment]
Would this bring risk that new networks are added or removed in the area and never submitted due to being scanned previously?

**Comment by mjaakko** [comment]
> Would this bring risk that new networks are added or removed in the area and never submitted due to being scanned previously?

No, there will be a time limit on how old reports will be used to check whether to stop scanning

**Comment by lgommans** [comment]
Speed would also be relevant to take into account here. In a train driving 50km/h for a while through switches/points near a station, I think I could go by a dozen times before having seen all the WiFis you might be able to receive there. If the app would see "ah I have a report near here from within the last week", it would miss probably >50% of the data even if you have scanning enabled for a month. Granted, from the rather isolated train (both in terms of track location and the train's walls / heat-repellent windows) it is exceptionally hard to receive nearby-ish networks and it's worth scanning a few more times than it would be at 50 km/h in a street, but still, from a car I would also want to scan several times, whereas with walking that's not necessary — if scanning isn't rate limited

I was still meaning to look into how many times I had to drive by a street until it finds no new networks, or what the maximum speed is where it still sees >90%. Would that be helpful here?

**Comment by DJCrashdummy** [comment]
> Speed would also be relevant to take into account here.

well... i guess it comes down to the fact, that **no measurements are perfect**: so - if i assumed correctly - having **more reports** for the same area resp. networks/hotspots/beacons - even from the same device - is helping to **improve** the over all **data quality**. ...so i'm not sure if that can be reasonably determined on a per device basis if the coverage of an area is already "sufficient" (and recent enough) or not.

at first i thought this feature would be awesome (in combination with #111), because then i would be able to simply have NeoStumbler on all the time and still haven't drained my battery all the time, but just "when it counts" aka. exploring not (currently) covered areas. but after thinking more about it, i'm not so sure about this feature any more. 😑

**Comment by lgommans** [comment]
> having more reports for the same area resp. networks/hotspots/beacons - even from the same device - is helping to improve the over all data quality. [...] after thinking more about it, i'm not so sure about this feature any more. 😑

No I think that's very fair. Maybe it could be approached more broadly, as "how to make battery drain such that it can be always-on" rather than "avoid covered locations". An option that doesn't involve location at all might be checking signals at a slow pace, perhaps whenever it sees ["significant motion"](https://developer.android.com/develop/sensors-and-location/sensors/sensors_motion#sensors-motion-significant), like see if there are WiFis, Bluetooth beacons, or cell towers it hasn't seen. The OS knows what tower you're on anyway, and one BT/WiFi scan per N minutes (e.g., 2) takes much less power than maintaining a GNSS fix, especially indoors where it searches for a while before confirming "ah yep, we're still at home, this is covered".

This would miss a few minutes of collection (on average: half of N) each time you enter a non-covered region, but if you return to the same place (same hotel, for example), it'll collect that street on the return journey. It might be better overall that people miss a small amount of data than that they have to manually check for coverage and battery and turn it on each time! And you can still turn it on manually, too.

> i'm not sure if that can be reasonably determined on a per device basis if the coverage of an area is already "sufficient"

For my idea to work across all devices, the server would have to implement an API which returns whether it knows all of the submitted beacons already (or, say, 95%), or adapt one of the APIs to include this information in the response. (The existing `/geosubmit` endpoint never returns info, and `/geolocate` would return success when it knows just 2 out e.g. 42 WiFis: Neostumbler would conclude "ok it knows where we are, this area must be covered".) Could check if such an addition is objectionable to implement for e.g. BeaconDB.

Otherwise it'll just be on-device and it doesn't need to have internet access for it to work either, but you do need to save scan history indefinitely or implement a bloom filter or so. This would then just save battery in areas where *you've* completely covered the data, which I bet is already 95% of the possible savings (work, home, your typical shops... how much time is a typical person travelling new places?). The only downside I can think of is that you're more likely to want to conserve battery when you're travelling to a new place, but you'd have that conflict of priorities no matter the solution
